shared/js/dataDriver.js

```diff
@@ -0,0 +1,283 @@
+          break;
+
+        case "put": {
+          // Ignore initial put from connecting.
+          let { data, path } = JSON.parse(event.data);
+          if (path === "/") {
+            return;
+          }
+
+          // Remove the leading "/" from the path.
+          this._processRecord(path.slice(1), data);
+          break;
+        }
+      }
+    }
+
+    /** ************* **
+     *** Data Source ***
+     ** ************* **/
+
+    /**
+     * Get the base url for the Firebase connection.
+     */
+    get BASE_FIREBASE() {
+      return "https://blinding-fire-8842.firebaseio.com/loop-test";
+    }
+
+    /**
+     * Get the maximum query limit value: 2^31 - 1.
+     */
+    get MAX_LIMIT() {
+      return 2147483647;
+    }
+
+    /**
+     * Get the server time for a local time or now.
+     */
+    getServerTime(time = Date.now()) {
+      return (this._clockSkew || 0) + time;
+    }
+
+    /**
+     * Generate an id based on the time.
+     */
+    makeId(time = this.getServerTime()) {
+      // Call the function that has locally scoped state.
+      return makeId(time);
+    }
+
+    /**
+     * Request the most recent chat messages between two times.
+     */
+    requestChat(startTime = 0, endTime = this.getServerTime(), limit = 0) {
+      let query = this._buildQuery({
+        orderBy: '"$key"',
+        // Fudge the time to get an inclusive range with the random parts.
+        startAt: `"chat!${this.makeId(Math.max(0, startTime - 1))}"`,
+        endAt: `"chat!${this.makeId(endTime + 1)}"`,
+        // Always have a limit to get things sorted correctly.
+        limitToLast: limit > 0 ? limit : this.MAX_LIMIT
+      });
+      return this._request("GET", this._buildUrl(), query);
+    }
+
+    /**
+     * Update a record additionally storing the timestamp.
+     */
+    update(type, id, value) {
+      let key = `${type}!${id}`;
+      return this._request("PUT", this._buildUrl(key), {
+        "timestamp": {
+          ".sv": "timestamp"
+        },
+        value: value
+      });
+    }
+
+    /** ****************** **
+     *** Internal Helpers ***
+     ** ****************** **/
+
+    /**
+     * Convert an object with key/value pairs into a query string.
+     */
+    _buildQuery(object) {
+      return Object.keys(object).map(key => `${key}=${object[key]}`).join("&");
+    }
+
+    /**
+     * Get a REST endpoint for a resource defaulting to the current room.
+     */
+    _buildUrl(resource = "", roomToken = this._roomToken) {
+      return `${this.BASE_FIREBASE}/${roomToken}/${resource}.json`;
+    }
+
+    /**
+     * Connect to a specified room for streaming updates and future requests.
+     */
+    _connectToRoom(token) {
+      this._roomToken = token;
+      this._sse = new EventSource(this._buildUrl() + "?" + this._buildQuery({
+        limitToLast: 1,
+        orderBy: '"timestamp"'
+      }));
+      this._sse.addEventListener("open", this);
+      this._sse.addEventListener("put", this);
```

Comment:
`addEventListener` takes an `EventListener` that has one method `handleEvent` but typically you see functions passed in because all javascript functions automatically implement the interface by calling the function. https://developer.mozilla.org/en-US/docs/Web/API/EventListener